From: Miguel Vadillo <vadillo@ti.com>
Date: Fri, 18 Mar 2011 22:37:55 +0000 (-0500)
Subject: syslink: ipu_pm: check if resources fifo is full
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~84
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=a193671881b8ede1e24efc1e1b9192571f810b53

syslink: ipu_pm: check if resources fifo is full

Check and notify if the software fifo got full and
reschedule the ipu_pm_work function to start attending
the req/rel resources.

If this happens the req/rel resource will be lost and
it is recommendable to increment the IPU_KFIFO_SIZE
that is currently 16.

Change-Id: If74c5d7769079da0361ecea23168253f28f843dc
Signed-off-by: Miguel Vadillo <vadillo@ti.com>
---

diff --git a/drivers/dsp/syslink/ipu_pm/ipu_pm.c b/drivers/dsp/syslink/ipu_pm/ipu_pm.c
index 22e01c8..d1c7f6b 100644
--- a/drivers/dsp/syslink/ipu_pm/ipu_pm.c
+++ b/drivers/dsp/syslink/ipu_pm/ipu_pm.c
@@ -640,6 +640,8 @@ static void ipu_pm_work(struct work_struct *work)
 	if (WARN_ON(params == NULL))
 		return;
 
+	pr_debug("Processing %d msgs\n", kfifo_len(&handle->fifo));
+
 	while (kfifo_len(&handle->fifo) >= sizeof(im)) {
 		/* set retval for each iteration asumming error */
 		retval = PM_UNSUPPORTED;
@@ -734,10 +736,11 @@ void ipu_pm_callback(u16 proc_id, u16 line_id, u32 event_id,
 	im.pm_msg = payload;
 
 	spin_lock_irq(&handle->lock);
-	if (kfifo_avail(&handle->fifo) >= sizeof(im)) {
+	if (kfifo_avail(&handle->fifo) >= sizeof(im))
 		kfifo_in(&handle->fifo, (unsigned char *)&im, sizeof(im));
-		queue_work(ipu_resources, &handle->work);
-	}
+	else
+		pr_err("fifo for resources full losing messages!\n");
+	queue_work(ipu_resources, &handle->work);
 	spin_unlock_irq(&handle->lock);
 }
 EXPORT_SYMBOL(ipu_pm_callback);
